### DISCUSSION TITLE
Bad UX for observing liquidity provision

### DISCUSSION THREAD

**Discussion by anna-carroll** [celo-org/optics-monorepo#830] [General]
## Problem
Right now, it is really bad UX to figure out whether a given Bridge Transfer message was pre-filled or not. It is possible, but it requires many steps and many RPC calls.

## Ideal Solution
Ideally, we would ship an upgrade to the `BridgeRouter` than adds a `PreFill` event when liquidity is provided. unfortunately, we'd have to first separate the `handle` logic out into its own contract in order to remain under the contract size limit

## Current Solution
Here's how to figure out if a given message was pre-filled or not, given the current UX:
**if the message HAS NOT been processed,**
- assemble the Pre-Fill ID for the message
- query the BridgeRouter contract for the `liquidityProvider`
- if it is address(0) that means no liquidity was provided; if it is not, then the address is the liquidity provider

**if the message HAS been processed,**
- query the chain for the Process event where the message was processed emitted by the Replica
- query the BridgeRouter to get the local address of the token based on the domain, ID pair for the token specified in the message
- query Transfer events emitted by the local token contract within the same transaction hash as the Process event
- Transfer will be from: BridgeRouter, to: recipient if it is a native token; OR from: null address, to: recipient if it is a non-native token; you can determine if the token is native by querying the BridgeRouter
- if recipient is NOT the recipient specified in the message, then the recipient is the liquidity provider; else if the recipient is the recipient specified in the message, then there was no liquidity provided
- if liquidity was provided, you can get the transaction hash / timestamp when liquidity was provided by looking for the Transfer event where to: recipient specified in the message, and amount: amount specified in message minus liquidity provider fee

**Comment by prestwich** [comment]
> Transfer amount will be full amount if liquidity was not provided OR full amount minus fee if liquidity was provided

Transfer amount in the message processing will ALWAYS be the full amount. the fee is handled by lowering the amount transferred in the pre-fill, not by reducing the eventual message processing

**Reply by anna-carroll** [reply]
yes yes you are totally right, good catch

**Reply by anna-carroll** [reply]
updated, thanks 🙏
